_posts/2024-09-05-digging-into-struct-initialization-performance.md

```diff
@@ -0,0 +1,251 @@
+   3cfcb:       e8 20 16 01 00          callq  4e5f0 <blake3_hasher_init>
+   3cfd0:       4c 89 ff                mov    %r15,%rdi
+   3cfd3:       48 89 de                mov    %rbx,%rsi
+   3cfd6:       4c 89 f2                mov    %r14,%rdx
+   3cfd9:       e8 02 17 01 00          callq  4e6e0 <blake3_hasher_update>
+   3cfde:       48 8d 5c 24 18          lea    0x18(%rsp),%rbx
+   3cfe3:       ba 20 00 00 00          mov    $0x20,%edx
+   3cfe8:       4c 89 ff                mov    %r15,%rdi
+   3cfeb:       48 89 de                mov    %rbx,%rsi
+   3cfee:       e8 3d 1f 01 00          callq  4ef30 <blake3_hasher_finalize>
+   3cff3:       c7 44 24 08 20 00 00    movl   $0x20,0x8(%rsp)
+   3cffa:       00
+   3cffb:       c6 44 24 0c 00          movb   $0x0,0xc(%rsp)
+   3d000:       48 89 5c 24 10          mov    %rbx,0x10(%rsp)
+   3d005:       48 8d 7c 24 08          lea    0x8(%rsp),%rdi
+   3d00a:       e8 41 fd ff ff          callq  3cd50 <*Slice(UInt8)@Slice(T)#hexstring:String>
+   3d00f:       48 81 c4 b0 07 00 00    add    $0x7b0,%rsp
+   3d016:       5b                      pop    %rbx
+   3d017:       41 5e                   pop    %r14
+   3d019:       41 5f                   pop    %r15
+   3d01b:       c3                      retq
+   3d01c:       0f 1f 40 00             nopl   0x0(%rax)
+```
+
+All the MOV instructions look like we are… copying the struct?
+
+I tried to allocate the struct on the stack and call an `init` method on it (it merely calls `#initialize` but we can’t call it directly because `#initialize` methods are protected). The only change in the `blake3` method is:
+
+```crystal
+hasher = uninitialized Blake3Hasher
+hasher.init
+```
+
+```
+SHA256 944.38k (  1.06µs) (± 0.97%)   225B/op   3.50× slower
+Blake3   3.30M (302.91ns) (± 2.24%)  80.0B/op        fastest
+```
+
+Performance is finally on par with calling the C functions directly. It means that LLVM is now doing its job at optimizing the abstraction away. Looking at the disassembly, the generated block is *identical* to the direct C function calls that I listed above. LLVM did a perfect job at optimizing the struct away!
+
+## What’s happening
+
+The struct is first initialized, then the 2KB are *copied* using too many assembly instructions to count them. Sure, it all happens on the stack, but it takes a awful lot of CPU time to copy all that, and it appears to do so *twice*? No wonder performance gets destroyed.
+
+Structs are initialized exactly like classes are: through constructor methods. While classes return a reference (one pointer) structs return the value itself that must be copied which can be an expensive operation. This is the origin of the problem. The Crystal codegen always generates a constructor method that looks like that:
+
+```crystal
+struct Blake3Hasher
+  def self.new(*args, **kwargs) : self
+    value = uninitialized self
+    value.initialize(*args, **kwargs)
+    value
+  end
+end
+```
+
+> [!TIP]
```

Comment:
**suggestion:** Make this an INFO callout. Or add `TIP` to the configuration.
```suggestion
> **INFO:** Tip
```

Currently supported callouts:
```console
$ yq e '.callouts' _config.yml
["INFO", "NOTE", "THANKS", "WARNING"]
```